account/payments.go

```diff
@@ -193,57 +194,84 @@ func (a *Service) sendPayment(paymentHash string, sendRequest *lnrpc.SendRequest
 /*
 AddInvoice encapsulate a given amount and description in a payment request
 */
-func (a *Service) AddInvoice(invoice *data.InvoiceMemo) (paymentRequest string, err error) {
+func (a *Service) AddInvoice(invoiceRequest *data.AddInvoiceRequest) (paymentRequest string, err error) {
 	lnclient := a.daemonAPI.APIClient()
 
 	// Format the standard invoice memo
+	invoice := invoiceRequest.InvoiceDetails
 	memo := formatTextMemo(invoice)
 
 	if invoice.Expiry <= 0 {
 		invoice.Expiry = defaultInvoiceExpiry
 	}
-	if err := a.waitReadyForPayment(); err != nil {
-		return "", err
-	}
-	channelsRes, err := lnclient.ListChannels(context.Background(), &lnrpc.ListChannelsRequest{
-		PrivateOnly: true,
-	})
+
+	maxReceive, _, _, err := a.getReceivePayLimit()
 	if err != nil {
+		a.log.Infof("failed to get account limits %v", err)
 		return "", err
 	}
+
+	// in case we don't need a new channel, we make sure the
+	// existing channels are active.
+	if maxReceive >= invoice.Amount {
+		if err := a.waitReadyForPayment(); err != nil {
+			return "", err
+		}
+	}
+
 	var routeHints []*lnrpc.RouteHint
-	for _, h := range channelsRes.Channels {
-		ci, err := lnclient.GetChanInfo(context.Background(), &lnrpc.ChanInfoRequest{
-			ChanId: h.ChanId,
-		})
-		if err != nil {
-			a.log.Errorf("Unable to add routing hint for channel %v error=%v", h.ChanId, err)
-			continue
+	maxReceiveMsat := maxReceive * 1000
+	amountMsat := invoice.Amount * 1000
+	smallAmountMsat := amountMsat
+	// We need the LSP to open a channel.
+	if maxReceiveMsat < amountMsat {
+		// In case we don't have enough capacity we need to request a channel.
+		lspInfo := invoiceRequest.LspInfo
+		if lspInfo == nil {
+			return "", errors.New("not enough capacity and missing LSP information")
 		}
-		remotePolicy := ci.Node1Policy
-		if ci.Node2Pub == h.RemotePubkey {
-			remotePolicy = ci.Node2Policy
+
+		// Calculate the channel fee.
+		amountForFeeMsat := float64(amountMsat - (lspInfo.ChannelFeeStartAmount * 1000))
+		channelFeeMsat := amountForFeeMsat * float64(lspInfo.ChannelFeeRate)
+		if channelFeeMsat < 0 {
+			channelFeeMsat = 0
 		}
-		a.log.Infof("adding routing hint = %v", h.RemotePubkey)
+
+		fakeChanID := &lnwire.ShortChannelID{BlockHeight: 1, TxIndex: 0, TxPosition: 0}
```

Comment:
I suggest perhaps to use BlockHeight: 1 >> 17 to be consistent with the zeroconf fake chanID

Reply:
My thought was actually to avoid this range so I won't need to check for conflicts with existing channels...